WordPress/src/main/java/org/wordpress/android/ui/stats/StatsWidgetProvider.java

```diff
@@ -0,0 +1,321 @@
+        String prevWidgetDataString = AppPrefs.getStatsWidgetData();
+        if (!StringUtils.isEmpty(prevWidgetDataString)) {
+            // we have prev data. Make sure it's fresh.
+            try {
+                JSONObject prevData = new JSONObject(prevWidgetDataString);
+                if (!prevData.has("date") || !prevData.getString("date").equals(currentDate)) {
+                    return null;
+                }
+
+                // make sure data is about the primary blog id. Same check is in the service.
+                if (!prevData.has("blog_id")  ||
+                        Integer.parseInt(prevData.getString("blog_id")) != primaryBlog.getRemoteBlogId()) {
+                    return null;
+                }
+
+                return prevData;
+            } catch (JSONException e) {
+                AppLog.e(AppLog.T.STATS, e);
+            }
+        }
+        return null;
+    }
+
+    /**
+     *  This is called when an instance the App Widget is created for the first time.
+     *  For example, if the user adds two instances of your App Widget, this is only called the first time.
+     */
+    @Override
+    public void onEnabled(Context context) {
+        AnalyticsTracker.track(AnalyticsTracker.Stat.STATS_WIDGET_ADDED);
+    }
+
+    /**
+     * This is called when the last instance of your App Widget is deleted from the App Widget host.
+     * This is where you should clean up any work done in onEnabled(Context), such as delete a temporary database.
+     * @param context The Context in which this receiver is running.
+     */
+    @Override
+    public void onDisabled(Context context) {
+        AnalyticsTracker.track(AnalyticsTracker.Stat.STATS_WIDGET_REMOVED);
+    }
+
+    /**
+     * This is called from the StatsService and from ApiHelper.RefreshBlogContentTask to check there is
+     * at least 1 Stats Widget installed for the current blog.
+     */
+    public static boolean shouldUpdateWidgetForBlog(Context context, String blogID) {
+        // Check if there are widgets installed on the device
+        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
+        ComponentName thisWidget = new ComponentName(context, StatsWidgetProvider.class);
+        if (appWidgetManager.getAppWidgetIds(thisWidget).length == 0) {
+            return false;
+        }
+
+        if (!AccountHelper.isSignedInWordPressDotCom()) {
+            AppLog.w(AppLog.T.STATS, "Not signed to WordPress.com.");
+            return false;
+        }
+
+        long parsedBlogID = Long.parseLong(blogID);
+        final long primaryBlogId = AccountHelper.getDefaultAccount().getPrimaryBlogId();
+        if (parsedBlogID != primaryBlogId) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public static void enqueueStatsRequestForBlog(Context context, String blogID, String date) {
+        // start service to get stats
+        Intent intent = new Intent(context, StatsService.class);
```

Comment:
Nothing in particular, the request is enqueued in the service. There is a check that ensures the an identical request is not already in the queue of the service. In that case the request is skipped, since the result of the previous enqueued request will be used.